src/src.jl

```diff
@@ -10,372 +10,159 @@
-    container.v[fsize, :] = pivot - q0
-    v = container.v
-
-    # compute the a_{fsize,i}, i< fsize
-    idx_list, not_empty = container.idx_list(fsize)
-
-    @fastmath for i in idx_list
-        container.a[fsize, i] = 0.0;
-        for j = 1:d
-            container.a[fsize, i] += v[i, j] * v[fsize, j]
-        end
-        container.a[fsize, i] *= 2 / z[i]
-    end
-    a = container.a
-
-    # update v_fsize to Q_fsize-\bar{Q}_fsize
-    for i in idx_list
-        for j=1:d
-            container.v[fsize, j] -= a[fsize, i] * v[i, j]
-        end
-    end
-
-    # compute z_fsize
-    container.z[fsize] = 0.0;
-    for j=1:d
-        container.z[fsize] += pow_2(container.v[fsize, j])
-    end
-    container.z[fsize] *= 2
-    z = container.z
-    # reject push if z_fsize too small
-    if z[fsize] < epsilon*squared_radius
-        return_val = false
-    else
-        # update c and arr_squared_r
-        e_var = -arr_squared_r[fsize-1]
-        for i=1:d
-            e_var += pow_2(pivot[i]-c[fsize-1, i])
-        end
-        container.f[fsize] = e_var / z[fsize]
-        f = container.f
-        for i=1:d
-            container.c[fsize, i] = c[fsize-1, i] + f[fsize] * v[fsize, i]
-        end
-        container.arr_squared_r[fsize] = arr_squared_r[fsize-1] + e_var * f[fsize] / 2
-    end
-    return_val
- end
-
-"""
-    new_center_and_radius!(pivot, container)
-
-Function for creating a miniball. All the math related
-for calculating the miniball can be found from the
-http://www.inf.ethz.ch/personal/gaertner/texts/own_work/esa99_final.pdf,
-from B. Gaertner, Fast and Robust Smallest Enclosing Balls, ESA 1999,
-from section 3. The Primitive Operation and after that.
-"""
-function new_center_and_radius!{F<:AbstractFloat}(pivot::Array{F, 1}, container::MBContainer)
-    dim = container.dim
-    return_value = true
-    fsize = container.fsize
+    center = container.center
 
     if fsize == 1
-        return_value = push_fsize_zero(pivot, container)
+        q0[:] = c[1,:] = pivot[:]
+        arr_squared_r[1] = 0.0
     else
-        return_value = push_fsize_not_zero(pivot, container)
-    end
+        return_val = true
+
+        # set v_fsize to Q_fsize
+        v[fsize, :] = pivot - q0
 
-    if return_value == true
-        for i=1:dim
-            container.center[i] = container.c[fsize, i]
+        # compute the a_{fsize,i}, if i < fsize
+        # update v_fsize to Q_fsize - \bar{Q}_fsize
+        for i=2:fsize-1
+            a[fsize, i] = 2.0 * dot(v[i, :], v[fsize, :]) / z[i]
+            v[fsize, :] -= a[fsize, i] * v[i, :]
         end
-        container.squared_radius = container.arr_squared_r[fsize]
-        container.fsize += 1
-        container.ssize = container.fsize
+
+        e_var = norm(pivot[:]-c[fsize-1, :])^2 - arr_squared_r[fsize-1]
+        z[fsize] = 2.0 * norm(v[fsize, :])^2
+        # compute z_fsize and reject push if z_fsize is too small
+        isapprox(z[fsize], 0.0) && return false

```

Comment:
`isapprox(x, 0.0)` is the same as `x == 0.0` in julia. You need to add some tolerance for it to becomue useful.
e.g `isapprox(x, 0.0, atol=eps(typeof(x))` or just`abs(x) < eps(...)`.

Reply:
In the original implementation, there is `pow_2(1e-21)`. I actually don't like this workaround at all, we should inspect in what geometric situations `f[size]` gets small and skip those situations in other parts of the code to make this stable. And assume that `f[size]` is big enough.

Reply:
IIRC there was no situation, where this is smaller then `eps(Float64)` in the tests. So you could add an assertion and see what happens.

Reply:
![image](https://user-images.githubusercontent.com/1264779/41787222-5c11b64e-7650-11e8-97b8-a4c7cbb81baf.png)
![image](https://user-images.githubusercontent.com/1264779/41787230-61d91694-7650-11e8-9c5f-dc950297bcf1.png)

I think it's some kind of least squares approximation or similar, and when z gets zero obviously the matrix is not invertible anymore. 

"Equation (8) shows that A_B may become ill-conditioned (and the entries of A_B^-1 very large and unreliable) if z evaluates to a very small number. The subsequent lemma develops a geometric interpretation of z from which we can see that this happens exactly if the new point q_m is very close to the affine hull of the previous ones. This can be the case e.g. if input points are very close together or even equal."

There is also a solution to this problem, but it looks that it's not implemented to the code? Anyway the code is probably going to explode is set two points with same coordinates or the difference between them is `eps`.

Reply:
At least we need a test to check that we get something reasonable when points are near each other. This is for #24.